libp2p/switch.nim

```diff
@@ -577,11 +617,27 @@ proc newSwitch*(peerInfo: PeerInfo,
       await conn.close()
     trace "Muxed connection done", conn
 
-  result.mount(identity)
+  switch.mount(identity)
   for key, val in muxers:
-    val.streamHandler = result.streamHandler
+    val.streamHandler = switch.streamHandler
     val.muxerHandler = proc(muxer: Muxer): Future[void] =
-      s.muxerHandler(muxer)
+      switch.muxerHandler(muxer)
+
+  proc peerEventHandler(peerId: PeerID, event: ConnEvent) {.async, gcsafe.} =
+    ## peer event handler
+    ##
+
+    if peerId notin switch.connManager:
+      trace "peer left", peerId
+      doAssert(event.kind == ConnEventKind.Disconnected)
+      await switch.triggerPeerEvents(peerId, PeerEvent.Left)
+    elif switch.connManager.connCount(peerId) == 1:
+      trace "peer joined", peerId
+      doAssert(event.kind == ConnEventKind.Connected)
+      await switch.triggerPeerEvents(peerId, PeerEvent.Joined)
+
+  switch.addConnEventHandler(peerEventHandler, ConnEventKind.Connected)
```

Comment:
won't this result in a strange order of events? if I register both a conneventhandler and a peereventhandler, I'd expect the peereventhandler to trigger before the conneventhandler (so I can create peer-related structures in there)

Reply:
and the reverse for disconnected, ie I'd expect all connections to be closed before peer.left triggers

Reply:
yeah, I'm not sure if this approach is going to work, but I was trying to prevent duplicating the logic in the connection events. I'll see if the order is going to be possible to guarantee and if not I'll rework it.

Reply:
> won't this result in a strange order of events? if I register both a conneventhandler and a peereventhandler, I'd expect the peereventhandler to trigger before the conneventhandler (so I can create peer-related structures in there)

Actually, I was thinking that we don't know of the peer unless the connection has finished upgrading and if we trigger the peer event before the connection event, we sort of violating this assumption, so it made sense to me to trigger the peer event after the first and last conn event have triggered. Maybe my assumption is wrong and it's ok to have the peer events trigger before and after the conn events?

Reply:
OK, so after some more consideration, I think the right order would be to trigger the `Joined` peer events before the first conn event and the `Left` after the last conn event for the peer. 

So the order is:

- Peer Joined
- Connected a
- Connected b
- ...
- Disconnected a
- Disconnected b
- Peer Left